File: MvcFormsVb/ViewModels/MVCViewModel.vb
Imports System
Imports System.ComponentModel
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Diagnostics
Imports System.Drawing
Imports System.Linq
Imports System.Reflection
Imports System.Text
Imports System.Windows.Forms
Imports Ssepan.Utility
Imports Ssepan.Application
Imports Ssepan.Application.MVC
Imports Ssepan.Application.WinForms
Imports Ssepan.Io
Imports MvcLibraryVb

'Namespace MvcFormsVb
''' <summary>
''' Note: this class can subclass the base without type parameters.
''' </summary>
Public Class MVCViewModel
    Inherits FormsViewModel(Of Bitmap, MVCSettings, MVCModel, MVCView)
#Region "Declarations"
#End Region

#Region "Constructors"
    'Note: not called, but need to be present to compile--SJS
    Public Sub New()
    End Sub

    Public Sub New _
    (
        propertyChangedEventHandlerDelegate As PropertyChangedEventHandler,
        actionIconImages As Dictionary(Of [String], Bitmap),
        settingsFileDialogInfo As FileDialogInfo
    )
        MyBase.New(propertyChangedEventHandlerDelegate, actionIconImages, settingsFileDialogInfo) '(In VB 2010, )VB caller cannot differentiate between members which differ only by an optional param--SJS
        Try
        Catch ex As Exception
            Log.Write(ex, MethodBase.GetCurrentMethod(), EventLogEntryType.[Error])
        End Try
    End Sub

    Public Sub New _
    (
        propertyChangedEventHandlerDelegate As PropertyChangedEventHandler,
        actionIconImages As Dictionary(Of [String], Bitmap),
        settingsFileDialogInfo As FileDialogInfo,
        view_ As MVCView
    )
        MyBase.New(propertyChangedEventHandlerDelegate, actionIconImages, settingsFileDialogInfo, view_) '(In VB 2010, )VB caller cannot differentiate between members which differ only by an optional param--SJS
        Try
        Catch ex As Exception
            Log.Write(ex, MethodBase.GetCurrentMethod(), EventLogEntryType.[Error])
        End Try
    End Sub
#End Region

#Region "Properties"
#End Region

#Region "Methods"
    ''' <summary>
    ''' model specific, not generic
    ''' </summary>
    Public Sub DoSomething()
        StatusMessage = [String].Empty
        ErrorMessage = [String].Empty

        Try
            '_actionIconImages["Xxx"],
            StartProgressBar("Doing something...", Nothing, Nothing, True, 33)

            ModelController(Of MVCModel).Model.SomeBoolean = Not ModelController(Of MVCModel).Model.SomeBoolean
            ModelController(Of MVCModel).Model.SomeInt += 1
            ModelController(Of MVCModel).Model.SomeString = DateTime.Now.ToString()

            ModelController(Of MVCModel).Model.SomeComponent.SomeOtherBoolean = Not ModelController(Of MVCModel).Model.SomeComponent.SomeOtherBoolean
            ModelController(Of MVCModel).Model.SomeComponent.SomeOtherInt += 1
            ModelController(Of MVCModel).Model.SomeComponent.SomeOtherString = DateTime.Now.ToString()

            ModelController(Of MVCModel).Model.StillAnotherComponent.StillAnotherBoolean = Not ModelController(Of MVCModel).Model.StillAnotherComponent.StillAnotherBoolean
            ModelController(Of MVCModel).Model.StillAnotherComponent.StillAnotherInt += 1
            ModelController(Of MVCModel).Model.StillAnotherComponent.StillAnotherString = DateTime.Now.ToString()

            UpdateStatusBarMessages(Nothing, Nothing, DateTime.Now.ToLongTimeString())

            ModelController(Of MVCModel).Model.Refresh()
        Catch ex As Exception
            Log.Write(ex, MethodBase.GetCurrentMethod(), EventLogEntryType.[Error])

            StopProgressBar(Nothing, [String].Format("{0}", ex.Message))
        Finally
            StopProgressBar("Did something.")
        End Try
    End Sub
#End Region

End Class
'End Namespace

File: MVCForms/ViewModels/MVCViewModel.cs
using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Windows.Forms;
using Ssepan.Utility;
using Ssepan.Application;
using Ssepan.Io;
using MVCLibrary;

namespace MVCForms
{
    /// <summary>
    /// Note: this class can subclass the base without type parameters.
    /// </summary>
    public class MVCViewModel :
        FormsViewModel<Bitmap, MVCSettings, MVCModel, MVCView>
    {
        #region Declarations

        #region Commands
        //public ICommand FileNewCommand { get; private set; }
        //public ICommand FileOpenCommand { get; private set; }
        //public ICommand FileSaveCommand { get; private set; }
        //public ICommand FileSaveAsCommand { get; private set; }
        //public ICommand FilePrintCommand { get; private set; }
        //public ICommand FileExitCommand { get; private set; }
        //public ICommand EditCopyToClipboardCommand { get; private set; }
        //public ICommand EditPropertiesCommand { get; private set; }
        //public ICommand ViewPreviousMonthCommand { get; private set; }
        //public ICommand ViewPreviousWeekCommand { get; private set; }
        //public ICommand ViewNextWeekCommand { get; private set; }
        //public ICommand ViewNextMonthCommand { get; private set; }
        //public ICommand HelpAboutCommand { get; private set; }
        #endregion Commands
        #endregion Declarations

        #region Constructors
        public MVCViewModel() { }//Note: not called, but need to be present to compile--SJS

        public MVCViewModel
        (
            PropertyChangedEventHandler propertyChangedEventHandlerDelegate,
            Dictionary<String, Bitmap> actionIconImages,
            FileDialogInfo settingsFileDialogInfo
        ) :
            base(propertyChangedEventHandlerDelegate, actionIconImages, settingsFileDialogInfo)
        {
            try
            {
            }
            catch (Exception ex)
            {
                Log.Write(ex, MethodBase.GetCurrentMethod(), EventLogEntryType.Error);
            }
        }
        #endregion Constructors

        #region Properties
        #endregion Properties

        #region Methods
        /// <summary>
        /// model specific, not generioc
        /// </summary>
        public void DoSomething()
        {
            StatusMessage = String.Empty;
            ErrorMessage = String.Empty;

            try
            {
                StartProgressBar
                (
                    "Doing something...",
                    null,
                    null, //_actionIconImages["Xxx"],
                    true,
                    33
                );

                ModelController<MVCModel>.Model.SomeBoolean = !ModelController<MVCModel>.Model.SomeBoolean;
                ModelController<MVCModel>.Model.SomeInt += 1;
                ModelController<MVCModel>.Model.SomeString = DateTime.Now.ToString();

                ModelController<MVCModel>.Model.SomeComponent.SomeOtherBoolean = !ModelController<MVCModel>.Model.SomeComponent.SomeOtherBoolean;
                ModelController<MVCModel>.Model.SomeComponent.SomeOtherInt += 1;
                ModelController<MVCModel>.Model.SomeComponent.SomeOtherString = DateTime.Now.ToString();

                //SettingsController<MVCSettings>.Settings.SomeBoolean = true;
                //SettingsController<MVCSettings>.Settings.SomeInt += 1;
                //SettingsController<MVCSettings>.Settings.SomeString = "test";

                UpdateStatusBarMessages(null, null, DateTime.Now.ToLongTimeString());

                ModelController<MVCModel>.Model.Refresh();
            }
            catch (Exception ex)
            {
                Log.Write(ex, MethodBase.GetCurrentMethod(), EventLogEntryType.Error);

                StopProgressBar(null, String.Format("{0}", ex.Message));
            }
            finally
            {
                StopProgressBar("Did something.");
            }
        }
        #endregion Methods

    }
}
